From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 18 Mar 2026 23:16:44 +0900
Subject: [PATCH] Disable Useless Particles

Original license: GPL v3
Original project: https://github.com/CraftCanvasMC/Canvas

It is recommended to enable these options, as the client displays most of these particles already, so the server-side particle logic is not needed

diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 6565b517710e6a2ed7a1fcf617b1a46279bab149..e117e1784cf777cd894a5a1fc9233ff9d20cf398 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -978,7 +978,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
 
         if (this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick
-        if (this.canSpawnSprintParticle()) {
+        if (!org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.disableUselessParticles && this.canSpawnSprintParticle()) { // Leaf - Disable Useless Particles
             this.spawnSprintParticle();
         }
 
@@ -2256,6 +2256,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
         float f1 = Mth.floor(this.getY());
 
+        if (!org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.disableUselessParticles) { // Leaf - Disable Useless Particles
         for (int i = 0; i < 1.0F + this.dimensions.width() * 20.0F; i++) {
             double d = (this.random.nextDouble() * 2.0 - 1.0) * this.dimensions.width();
             double d1 = (this.random.nextDouble() * 2.0 - 1.0) * this.dimensions.width();
@@ -2276,6 +2277,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             double d1 = (this.random.nextDouble() * 2.0 - 1.0) * this.dimensions.width();
             this.level().addParticle(ParticleTypes.SPLASH, this.getX() + d, f1 + 1.0F, this.getZ() + d1, deltaMovement.x, deltaMovement.y, deltaMovement.z);
         }
+        } // Leaf - Disable Useless Particles
 
         this.gameEvent(GameEvent.SPLASH);
     }
@@ -4022,6 +4024,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     protected static void sendBubbleColumnParticles(Level level, BlockPos pos) {
+        if (org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.disableUselessParticles) return; // Leaf - Disable Useless Particles
         if (level instanceof ServerLevel serverLevel) {
             for (int i = 0; i < 2; i++) {
                 serverLevel.sendParticles(
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 0ad86184b239b43fde986ba8ccd94d2dd82aa628..b7392e03d05a9d3ddb5d31d79ce47955a3aa7728 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1008,7 +1008,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             List<ParticleOptions> list = this.entityData.get(DATA_EFFECT_PARTICLES);
-            if (!list.isEmpty()) {
+            if (!org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.disableUselessParticles && !list.isEmpty()) { // Leaf - Disable Useless Particles
                 boolean flag = this.entityData.get(DATA_EFFECT_AMBIENCE_ID);
                 int i = this.isInvisible() ? 15 : 4;
                 int i1 = flag ? 5 : 1;
@@ -2873,7 +2873,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 this.swapHandItems();
                 break;
             case 60:
-                this.makePoofParticles();
+                if (!org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.disableUselessParticles) this.makePoofParticles(); // Leaf - Disable Useless Particles
                 break;
             case 65:
                 this.breakItem(this.getItemBySlot(EquipmentSlot.BODY));
